### PR TITLE
Upgrade `fuse_rot_angles` to preserve derivatives and global phases

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
The current implementation of `fuse_rot_angles`, which is used by `merge_rotations` and `single_qubit_fusion`, has two issues:
1. It does not necessarily preserve global phases. As we move towards global-phase aware standards, this becomes an issue where it wasn't one before.
2. Its derivative is wrong, forming part (but not all) of the bug #5715. In particular, the custom handling of special input values prevents the calculation of correct derivatives, and `fuse_rot_angles` at singular points leads to wrong derivatives, rather than NaN values, which are mathematically well-motivated.
3. A minor technical issue is that the implementation requires nested conditionals, leading to a good bit of code and separate handling of traced JAX code, making it more complex.

**Description of the Change:**
The implementation of `fuse_rot_angles` is remade entirely. 

**Benefits:**
The remade code uses a comparably simple mathematical expression to compute the fused rotation angles that
1. preserves global phases
2. has the correct derivative everywhere except for well-understandable, predictable singular points. These singular points make sense because rotation fusion is not a smooth map everywhere. The predictability allowed us to write a dedicated test that confirms our understanding of the singularities, at least within a large set of special test points.
3. does not require conditionals beyond those that are implemented in `qml.math.arctan2` anyways, and thus available in all ML interfaces, including JAX with JIT-compatibility.

In summary, the global phases are fixed, Jacobians are only ever NaNs, rather than wrong, and Jacobians are only NaNs when they were NaN or wrong in the current implementation.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
#5715 (not fixed entirely, just partially)

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -33,6 +33,9 @@
 
 <h3>Improvements 🛠</h3>
 
+* `fuse_rot_angles` respects now the global phase of the combined rotations.
```
Comment (vincentmr) [inline doc/releases/changelog-dev.md:36]: Nitpick.
```suggestion
* `fuse_rot_angles` now respects the global phase of the combined rotations.
```

File: pennylane/transforms/optimization/merge_rotations.py [hunk 2]
```diff
@@ -70,10 +70,29 @@ def circuit(x, y, z):
...
     0.9553364891256055
 
+    .. details::
+        :title: Details on merging ``Rot`` gates
+        :href: details-on-rot
+
+        When merging two :class:`~.pennylane.Rot` gates, there are a number of details to consider:
+
+        First, the output angles are not always uniquely defined, because Euler angles are not
+        defined uniquely for some rotations. ``merge_rotations`` makes a particular choice in
```
Comment (vincentmr) [inline pennylane/transforms/optimization/merge_rotations.py:80]:
```suggestion
        unique for some rotations. ``merge_rotations`` makes a particular choice in
```

File: pennylane/transforms/optimization/single_qubit_fusion.py [hunk 3]
```diff
@@ -36,7 +36,7 @@ def single_qubit_fusion(
     (on the same qubit) with that property defined will be fused into one ``Rot``.
 
     Args:
-        tape (QNode or QuantumTape or Callable): A quantum circuit.
+        tape (QNode or QuantumTape or callable): A quantum circuit.
```
Comment (vincentmr) [inline pennylane/transforms/optimization/single_qubit_fusion.py:39]: There seem to be more `Callable` than `callable` in the code base. Is one correct the other not? Are they just different?

File: pennylane/transforms/optimization/single_qubit_fusion.py [hunk 4]
```diff
@@ -96,6 +113,116 @@ def qfunc(r1, r2):
...
+            &= \sqrt{c_1^2 c_2^2 + s_1^2 s_2^2 - 2 c_1 c_2 s_1 s_2 \cos(\omega_1 + \phi_2)}.
+
+        Now we again make a choice and pick :math:`\theta_f` to be non-negative:
+
+        .. math::
+
+            \theta_f = 2\arccos(|x|).
+
+        We can extract the angle combination :math:`\alpha_f` from :math:`x` above as well via
+        :math:`\operatorname{arg}(x)`, which can be readily computed with :math:`arctan`:
```
Comment (vincentmr) [inline pennylane/transforms/optimization/single_qubit_fusion.py:163]:
```suggestion
        :math:`\operatorname{arg}(x)`, which can be readily computed with :math:`\arctan`:
```

File: pennylane/transforms/optimization/single_qubit_fusion.py [hunk 5]
```diff
@@ -96,6 +113,116 @@ def qfunc(r1, r2):
...
+            \phi_f, \omega_f \in [-\pi, \pi],\quad \theta_f \in [0, \pi].
+
+        Close to the boundaries of these intervals, ``single_qubit_fusion`` exhibits
+        discontinuities, depending on the combination of input angles.
+        These discontinuities also lead to singular (non-differentiable) points as discussed below.
+
+        **Differentiability:**
+
+        The function derived above is differentiable almost everywhere.
+        In particular, there are two problematic scenarios at which the derivative is not defined:
```
Comment (vincentmr) [inline pennylane/transforms/optimization/single_qubit_fusion.py:221]:
```suggestion
        In particular, there are two problematic scenarios at which the derivative is not defined.
```

File: pennylane/transforms/optimization/optimization_utils.py [hunk 6]
```diff
@@ -47,146 +41,105 @@ def find_next_gate(wires, op_list):
...
+    angles_1 = qml.math.asarray(angles_1)
+    angles_2 = qml.math.asarray(angles_2)
+    # This sum is only computed to obtain a dtype-coerced object that respects
+    # TensorFlow's coercion rules between Python/NumPy objects and TF objects.
+    _sum = angles_1 + angles_2
+    # moveaxis required for batched inputs
+    phi1, theta1, omega1 = qml.math.moveaxis(qml.math.cast_like(angles_1, _sum), -1, 0)
+    phi2, theta2, omega2 = qml.math.moveaxis(qml.math.cast_like(angles_2, _sum), -1, 0)
+
+    if qml.math.allclose(omega1 + phi2, 0.0):
```
Comment (vincentmr) [inline pennylane/transforms/optimization/optimization_utils.py:56]: Should the following logic also take care of modulo-pi-ness (i.e. special cases when angles are close to +/- N * pi), or is it taken care of somewhere else or not applicable?

File: pennylane/transforms/optimization/optimization_utils.py [hunk 7]
```diff
@@ -47,146 +41,105 @@ def find_next_gate(wires, op_list):
...
+        return qml.math.stack([phi1, theta1 + theta2, omega2])
```
Comment (vincentmr) [inline pennylane/transforms/optimization/optimization_utils.py:57]: Should we deal with `theta1 + theta2 > pi` here?

File: tests/transforms/test_optimization/test_optimization_utils.py [hunk 8]
```diff
@@ -56,78 +52,199 @@ def test_find_next_gate(self, wires, op_list, next_gate_idx):
...
 
-        def original_ops():
-            qml.Rot(*angles_1, wires=0)
-            qml.Rot(*angles_2, wires=0)
+        angles_1, angles_2 = qml.numpy.array(angles_1), qml.numpy.array(angles_1)
+        matrix_expected = self.mat_from_prod(angles_1, angles_2)
+        fused_angles = fuse_rot_angles(angles_1, angles_2)
+        matrix_obtained = qml.Rot(*fused_angles, wires=0).matrix()
+
+        assert qml.math.allclose(matrix_expected, matrix_obtained)
```
Comment (vincentmr) [inline tests/transforms/test_optimization/test_optimization_utils.py:133]: Can we bundle these lines into a function to avoid duplication below?

**Review by vincentmr — COMMENTED** [review]
Nice work @dwierichs . I have a couple questions, but this looks ready to merge to me :)

File: pennylane/transforms/optimization/optimization_utils.py (hunk 6 above)
Comment (dwierichs) [reply pennylane/transforms/optimization/optimization_utils.py:56]: That's a fair question! The previous version of this function did not have any guarantees for the output range, do we want to introduce this guarantee? :thinking: I'd hope that it is not a problem if all ML interfaces define the derivative of `mod` everywhere :grimacing:

File: pennylane/transforms/optimization/optimization_utils.py (hunk 7 above)
Comment (dwierichs) [reply pennylane/transforms/optimization/optimization_utils.py:57]: See above. Note that `phi1` alone or `omega2` could also be larger than `pi` and we would not modify it here.

File: pennylane/transforms/optimization/single_qubit_fusion.py (hunk 3 above)
Comment (dwierichs) [reply pennylane/transforms/optimization/single_qubit_fusion.py:39]: Ah, I actually don't know, this is a bit of a leftover :D I'll just revert to stick to the majority vote.

Comment (dwierichs) [inline pennylane/transforms/optimization/single_qubit_fusion.py:39]:
```suggestion
        tape (QNode or QuantumTape or Callable): A quantum circuit.
```

File: pennylane/transforms/optimization/single_qubit_fusion.py [hunk 9]
```diff
@@ -45,7 +45,8 @@ def single_qubit_fusion(
             be fused will be fused.
 
     Returns:
-        qnode (QNode) or quantum function (Callable) or tuple[List[QuantumTape], function]: The transformed circuit as described in :func:`qml.transform <pennylane.transform>`.
+        qnode (QNode) or quantum function (callable) or tuple[List[QuantumTape], callable]:
```
Comment (dwierichs) [inline pennylane/transforms/optimization/single_qubit_fusion.py:48]:
```suggestion
        qnode (QNode) or quantum function (Callable) or tuple[List[QuantumTape], Callable]:
```

File: tests/transforms/test_optimization/test_optimization_utils.py (hunk 8 above)
Comment (dwierichs) [reply tests/transforms/test_optimization/test_optimization_utils.py]: Done, worked very well!

File: pennylane/transforms/optimization/optimization_utils.py (hunk 6 above)
Comment (vincentmr) [reply pennylane/transforms/optimization/optimization_utils.py:54]: I have nothing specific in mind, I'm just wondering if there are other special cases away from zero which we could handle here. For instance, I think `omega`, `phi` are taken in [-pi, pi]. If they both are almost pi, should we not merge them and also return `qml.math.stack([phi1, theta1 + theta2, omega2])`? Not a blocker from me in any case.

Comment (dwierichs) [reply pennylane/transforms/optimization/optimization_utils.py:54]: Ah, I understand! Sorry, I thought you were thinking of doing `np.mod(theta1+theta2, 2*np.pi)` or something like that.

We could, in principle, do the following:
```python
if qml.math.allclose(qml.math.mod(omega1 + phi2+0.1, 2* np.pi)-0.1, 0.0):
    return qml.math.stack([phi1 + omega1 + phi2, theta1 + theta2, omega2])
```
If they are zero, taken together, this doesn't change things, and if they are other multiples of 2*pi, they lead to either a global phase or the identity, which can be pulled up front.

Similarly, we could think about pulling factors of pi in the `RY` part through.

I'm happy to include these things, but I'm not sure how much optimization we want :)

CC @trbromley do you have an opinion here?

File: pennylane/transforms/optimization/optimization_utils.py (hunk 7 above)
Comment (dwierichs) [reply pennylane/transforms/optimization/optimization_utils.py:55]: Okay, nevermind, this is actually a different comment, right? :) 
I think I would not do modulus here, and live with the potential outputs outside of `[-pi,pi]`.

Comment (vincentmr) [reply pennylane/transforms/optimization/optimization_utils.py:55]: > different comment, right?

Yes.

File: pennylane/transforms/optimization/optimization_utils.py (hunk 6 above)
Comment (vincentmr) [reply pennylane/transforms/optimization/optimization_utils.py:54]: > I thought you were thinking of doing `np.mod(theta1+theta2, 2*np.pi)` or something like that.

That too :D it was a bit left to the imagination. But first and foremost, I was thinking introducing this kind of logic `qml.math.allclose(qml.math.mod(omega1 + phi2+0.1, 2* np.pi)-0.1, 0.0)`. Let's see what @trbromley thinks.

Comment (trbromley) [reply pennylane/transforms/optimization/optimization_utils.py:54]: Overall, I think our broad preference is towards:
- Code correctness, i.e., incorrect numeric answers are not provided.
- Avoiding over-optimized simplifications during higher-level Python passes. Going forward (outside the scope of this PR), we can look to add low-level compiler passes that could work across the ecosystem (Catalyst and Python PennyLane). Not sure exactly what this looks like right now, it's perhaps linked to how we decide to get transforms and program capture to work together.

For this PR, I'd recommend we focus on code correctness, but leave it up to you for the particular decision here.

Comment (dwierichs) [reply pennylane/transforms/optimization/optimization_utils.py:54]: In this case I agree to go for correctness and not introduce this change here.

File: pennylane/transforms/optimization/optimization_utils.py [hunk 10]
```diff
@@ -47,146 +41,105 @@ def find_next_gate(wires, op_list):
...
-def _singular_quat_to_zyz(qw, qx, qy, qz, y_arg, abstract_jax=False):
-    """Compute the ZYZ angles for the singular case of qx = qy = 0"""
-    # pylint: disable=too-many-arguments
-    z1_arg1 = 2 * (qx * qy + qz * qw)
-    z1_arg2 = 1 - 2 * (qx**2 + qz**2)
+def _try_no_fuse(angles_1, angles_2):
+    """Try to combine rotation angles without trigonometric identities
+    if some angles in the input angles vanish."""
+    # This sum is only computed to obtain a dtype-coerced object that respects
+    # TensorFlow's coercion rules between Python/NumPy objects and TF objects.
```
Comment (Qottmann) [inline pennylane/transforms/optimization/optimization_utils.py:48]: cool trick :)

File: pennylane/transforms/optimization/single_qubit_fusion.py [hunk 11]
```diff
@@ -96,6 +113,132 @@ def qfunc(r1, r2):
...
+        :math:`\operatorname{arg}(x)`, which can be readily computed with :math:`\arctan`:
+
+        .. math::
+
+            \alpha_f = -\arctan\left(
+            \frac{-c_1c_2\sin(\alpha_1+\alpha_2)-s_1s_2\sin(\beta_2-\beta_1)}
+            {c_1c_2\cos(\alpha_1+\alpha_2)-s_1s_2\cos(\beta_2-\beta_1)}
+            \right).
+
+        We can use the standard numerical function :math:`\operatorname{arctan2}`, which
```
Comment (Qottmann) [inline pennylane/transforms/optimization/single_qubit_fusion.py:172]:
```suggestion
        We can use the standard numerical function ``arctan2``, which
```

optional

**Review by vincentmr — APPROVED** [review]
LGTM, nice one @dwierichs !

File: pennylane/transforms/optimization/merge_rotations.py [hunk 12]
```diff
@@ -70,10 +70,29 @@ def circuit(x, y, z):
...
+        When merging two :class:`~.pennylane.Rot` gates, there are a number of details to consider:
+
+        First, the output angles are not always defined uniquely, because Euler angles are not
+        unique for some rotations. ``merge_rotations`` makes a particular choice in
+        this case.
+
+        Second, ``merge_rotations`` is not differentiable everywhere when used on ``Rot``.
+        It has singularities for specific rotation angles where the derivative will be NaN.
+
+        Finally, this function can be numerically unstable at singular points.
```
Comment (trbromley) [inline pennylane/transforms/optimization/merge_rotations.py:86]:
```suggestion
        Finally, this function can be numerically unstable near singular points.
```

File: tests/transforms/test_optimization/test_optimization_utils.py [hunk 13]
```diff
@@ -56,78 +52,173 @@ def test_find_next_gate(self, wires, op_list, next_gate_idx):
...
+        ([0.0, 0.0, 0.05], [0.0, 0.0, -0.05]),
+        ([0.05, 0.2, 0.0], [0.0, -0.6, 0.12]),
+        ([0.05, -1.34, 4.12], [0.0, 0.2, 0.12]),
+        ([0.05, -1.34, 4.12], [0.3, 0.0, 0.12]),
+        ([np.pi, np.pi / 2, 0.0], [0.0, -np.pi / 2, 0.0]),
+        ([0.9, np.pi / 2, 0.0], [0.0, -np.pi / 2, 0.0]),
+        ([0.9, np.pi / 2, np.pi / 2], [-np.pi / 2, -np.pi / 2, 0.0]),
+    ]
+
+    def run_interface_test(self, angles_1, angles_2):
```
Comment (Qottmann) [inline tests/transforms/test_optimization/test_optimization_utils.py:78]: thats pretty neat!

**Review by Qottmann — COMMENTED** [review]
Had a look at the tests now as well. Looks good so far, but aren't we missing tests for differentiating with autograd, torch and tf?

**Comment by dwierichs** [comment]
>aren't we missing tests for differentiating with autograd, torch and tf?

Yes, you're right. I was happy to include derivatives in the tests at all, and noticed they started taking quite some time already. But I can add the other interfaces as well :+1:

**Comment by dwierichs** [comment]
I'm working on extending the tests to other interfaces, and the tests start hitting O(10) seconds for this function. 
Previously the coverage was much worse, but I'm not sure how many resources we want to throw at this?

**Comment by dwierichs** [comment]
I think it is fine now for all interfaces, but I can't get TensorFlow to work with broadcasting and the non-broadcasted version takes ages. @Qottmann would it be fine to skip the TensorFlow interface :grimacing:

**Comment by trbromley** [comment]
> I think it is fine now for all interfaces, but I can't get TensorFlow to work with broadcasting and the non-broadcasted version takes ages. @Qottmann would it be fine to skip the TensorFlow interface 😬

TensorFlow is always being a pain :cry:

In general we're happy for some things to slip in terms of feature parity for TensorFlow. But we want to be avoiding giving wrong answers - would that be the case here?

**Comment by dwierichs** [comment]
So I think the code should in principle work with full features also with Tensorflow, but I can't seem to make it work with TF's `vectorized_map` in the test. This is more of a feature disparity of my TensorFlow skills, rather than PennyLane's feature support :D

However, other interfaces like Autograd also  need to iterate manually within the test, instead of using something like JAX's `vmap`, making them very slow.
In the end, the question is whether we want to pay more than a minute (!) of testing time to this single function, just to guarantee correctly valued derivatives at special points?

**Review by Qottmann — APPROVED** [review]
Awesome job @dwierichs !

That was an impressive upgrade in test coverage, which makes me almost wonder if this shooting over the top 😅 For now, happy with it as it covers all 👍
